93 camunda form data props

MaxTru opened the pull request:
closes #93
requires bpmn-io/properties-panel#93

![formData](https://user-images.githubusercontent.com/42800119/133290759-ecac68d9-6b70-44e4-bcf8-0591bab9b524.gif)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/bpmn-io/bpmn-properties-panel?pullRequest=97) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/bpmn-io/bpmn-properties-panel?pullRequest=97) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/bpmn-io/bpmn-properties-panel?pullRequest=97) it.</sub>

MaxTru commented:
Now since nested lists were merged, this is ready for review

pinussilvestrus commented on src/provider/camunda-platform/properties/FormField.js:
```diff
@@ -0,0 +1,487 @@
+
+  const constraints = (validation && validation.get('constraints')) || [];
+
+  function Constraint(constraint, index) {
+    return (
+      <CollapsibleEntry
+        id={ `${id}-constraint-${index}` }
+        entries={ FormFieldConstraint({ idPrefix: `${id}-field-${index}`, element, constraint }) }
+        label={ constraint.get('name') || '<empty>' }
+      />
+    );
+  }
```
We can achieve the autoOpen + autoFocus behavior here by using the third parameter of the `renderItem` function.

```js
function Constraint(constraint, index, open) {
  return (
    <CollapsibleEntry
      id={ `${id}-constraint-${index}` }
      entries={ FormFieldConstraint({ idPrefix: `${id}-field-${index}`, element, constraint }) }
      label={ constraint.get('name') || '<empty>' }
      open={ open }
    />
  );
}
  ```

MaxTru replied:
Good point! Thanks. I changed that.

Note: I also added the sorting (to be in line with our default behavior) for constraints and properties (but not for values of an enum, since IMO the user wants to understand which values are shown in which order). cc @andreasgeier

pinussilvestrus commented on src/provider/camunda-platform/properties/FormField.js:
```diff
@@ -0,0 +1,487 @@
+      <CollapsibleEntry
+        id={ `${id}-constraint-${index}` }
+        entries={ FormFieldConstraint({ idPrefix: `${id}-field-${index}`, element, constraint }) }
+        label={ constraint.get('name') || '<empty>' }
```
```suggestion
        label={ constraint.get('name') || translate('<empty>') }
```

MaxTru replied:
good catch, ty

pinussilvestrus commented on src/provider/camunda-platform/properties/FormField.js:
```diff
@@ -0,0 +1,487 @@
+  function addConstraint() {
+    const commands = [];
+
+    // 1. ensure validation
```
Just a minimal thing, but I tend to use (1), (2), ... for step-by-step comments.

MaxTru replied:
being consistent makes sense - ty

pinussilvestrus commented on src/provider/camunda-platform/properties/FormField.js:
```diff
@@ -0,0 +1,487 @@
+      <CollapsibleEntry
+        id={ `${id}-constraint-${index}` }
+        entries={ FormFieldProperty({ idPrefix: `${id}-field-${index}`, element, property }) }
+        label={ property.get('id') || '<empty>' }
```
```suggestion
        label={ property.get('id') || translate('<empty>' )}
```

MaxTru replied:
good catch - ty

pinussilvestrus commented on src/provider/camunda-platform/properties/FormField.js:
```diff
@@ -0,0 +1,487 @@
+
+  const propertyEntries = (properties && properties.get('values')) || [];
+
+  function Property(property, index) {
```
Same as mentioned in https://github.com/bpmn-io/bpmn-properties-panel/pull/97/files#r707995735

pinussilvestrus commented on src/provider/camunda-platform/properties/FormDataProps.js:
```diff
@@ -0,0 +1,237 @@
+  const formFields = getFormFieldsList(element) || [];
+
+  const items = formFields.map((formField, index) => {
+    const id = 'FormField-' + index;
```
We need to make sure the ids are unique per element (as we do in [other lists](https://github.com/bpmn-io/bpmn-properties-panel/blob/main/src/provider/zeebe/properties/InputProps.js#L41)). Otherwise, the item states got persisted across different elements.

![Kapture 2021-09-14 at 12 07 48](https://user-images.githubusercontent.com/9433996/133238907-711992a6-873b-4084-b71f-5d6180a8d7ba.gif)

MaxTru replied:
sure, thanks - adjusted

pinussilvestrus commented on test/spec/provider/camunda-platform/FormField.spec.js:
```diff
@@ -0,0 +1,984 @@
+    }));
+
+
+    it('should not display empty option given defined type', inject(async function(elementRegistry, selection) {
```
Asking again from above: Is there a special reason why we only show this empty option once?

MaxTru replied:
See https://github.com/bpmn-io/bpmn-properties-panel/pull/97#discussion_r708238987

pinussilvestrus reviewed (commented):
Quite a change! Some comments attached above 👍

MaxTru commented on src/provider/camunda-platform/properties/FormDataProps.js:
```diff
@@ -0,0 +1,237 @@
+
+      // remove camunda:formData if there are no formFields anymore
+      if (formFields.length === 1) {
+        commands.push({
```
True, but following our latest hour of code discussion, let's keep it as-is for now and potentially improve once we have defined the target state

MaxTru commented on src/provider/camunda-platform/properties/FormField.js:
```diff
@@ -0,0 +1,487 @@
+
+    // For the initial state only, we want to show an empty state
+    if (formField.get('type') === undefined) {
+      options.unshift({ label: '', value: '' });
```
> Shouldn't that have the label <none> then?
True, I updated that. Thanks!

>And why can't we have it all the time?
It then conflicts the <custom type>. If you have a <custom type> and delete all characters (so empty input), the two select options match and the state of the select can not clearly be defined. 
Therefore I decided to only show it initially: (1) makes implementation much easier, (2) from a business perspective from my understanding it makes no harm (you must have a type anyhow).

Makes sense?

pinussilvestrus replied:
Makes sense!

MaxTru commented on src/provider/camunda-platform/properties/FormField.js:
```diff
@@ -0,0 +1,487 @@
+
+  const getOptions = () => {
+    const options = [
+      { label: translate('string'), value: 'string' },
+      { label: translate('long'), value: 'long' },
+      { label: translate('boolean'), value: 'boolean' },
+      { label: translate('date'), value: 'date' },
+      { label: translate('enum'), value: 'enum' },
+      { label: translate('<custom type>'), value: CUSTOM_TYPE_VALUE },
```
Good point, will do. Ty

philippfromme commented:
Group titles are title case, right?

![image](https://user-images.githubusercontent.com/7633572/133391677-34f76b09-f8ae-443c-9c1f-817d689ce6eb.png)

MaxTru commented:
> Group titles are title case, right?
> 
> ![image](https://user-images.githubusercontent.com/7633572/133391677-34f76b09-f8ae-443c-9c1f-817d689ce6eb.png)

No, sentence case - but we are not yet consistent XD cc. @andreasgeier

philippfromme commented:
Adding a field doesn't expand the group. Am I missing something?

![KBXVjubMi8](https://user-images.githubusercontent.com/7633572/133392127-104768eb-0204-49e8-b44a-6b3f05a053d9.gif)

philippfromme commented:
> No, sentence case - but we are not yet consistent XD cc. @andreasgeier

Can we achieve this before the release?

MaxTru commented:
> Adding a field doesn't expand the group. Am I missing something?
> 
> ![KBXVjubMi8](https://user-images.githubusercontent.com/7633572/133392127-104768eb-0204-49e8-b44a-6b3f05a053d9.gif)

That is right. But currently we cannot change that. The reason for that is:
* We don't want the list to be auto-sorted (the sorting should follow the XML, this is important for formFields)
* However, disabling the auto-sort, also disables the auto-expand
* The current list implementation does not allow to configure these things independetly from each other (see https://docs.google.com/document/d/1k2rBbbIvReLOgxaAZrUHqraaAHJh6yWqjV8FABd-Ylg/edit)
* Since generated formData is anyhow not our preferred way of using forms, my assumption is that we agreed, that this is good enough. 

cc. @andreasgeier @pinussilvestrus  please correct me if wrong

philippfromme commented on src/provider/camunda-platform/properties/FormField.js:
```diff
@@ -0,0 +1,525 @@
+      { label: translate('<custom type>'), value: CUSTOM_TYPE_VALUE },
+    ];
+
+    // For the initial state only, we want to show an empty state
```
```suggestion
    // for the initial state only, we want to show an empty state
```

philippfromme commented on src/provider/camunda-platform/properties/FormField.js:
```diff
@@ -0,0 +1,525 @@
+      { label: translate('enum'), value: 'enum' },
+      { label: translate('long'), value: 'long' },
+      { label: translate('string'), value: 'string' },
+      { label: translate('<custom type>'), value: CUSTOM_TYPE_VALUE },
```
```suggestion
      { label: translate('<custom type>'), value: CUSTOM_TYPE_VALUE }
```

MaxTru replied:
ty, corrected

andreasgeier commented:
> Since generated formData is anyhow not our preferred way of using forms, my assumption is that we agreed, that this is good enough.
> 
> cc. @andreasgeier @pinussilvestrus please correct me if wrong

I disagree. 

This is also an issue for other lists, at least for errors, too.

We discussed and agreed that for now, we will not place a hint in the UI to inform the user that this is a list where order matters and the sort is disabled. But I expect the same interaction for all our list sections including that the create event (1) opens the sections, (2) opens the created item, (3) and focuses the first entry (which defines the list item name).

@MaxTru I would propose to discuss how we solve this. Maybe as an additional issue outside of this PR, but mandatory from my perspective.

andreasgeier commented:
> > No, sentence case - but we are not yet consistent XD cc. @andreasgeier
> 
> Can we achieve this before the release?

We plan it as part of the design finishing. Not yet sure what is feasible until the Oct release but of course we want to align it.

philippfromme commented on src/provider/camunda-platform/properties/FormDataProps.js:
```diff
@@ -0,0 +1,237 @@
+      const parent = extensionElements;
+
+      formData = createElement('camunda:FormData', {
+        values: []
```
`FormData` has no property called `values`. I think, this should be `fields` instead. (cf. https://github.com/bpmn-io/bpmn-js-properties-panel/blob/master/lib/provider/camunda/parts/FormProps.js#L148).

Currently the exported XML looks like this:

```xml
<userTask id="Task_1">
  <extensionElements>
    <camunda:formData values="">
      <camunda:formField label="Foo" />
    </camunda:formData>
  </extensionElements>
</userTask>
```

MaxTru replied:
Good catch, thanks. Indeed it is `fields` (see [moddle](https://github.com/camunda/camunda-bpmn-moddle/blob/master/resources/camunda.json#L1018))

Is adjusted

MaxTru commented:
> > Since generated formData is anyhow not our preferred way of using forms, my assumption is that we agreed, that this is good enough.
> > cc. @andreasgeier @pinussilvestrus please correct me if wrong
> 
> I disagree.
> 
> This is also an issue for other lists, at least for errors, too.
> 
> We discussed and agreed that for now, we will not place a hint in the UI to inform the user that this is a list where order matters and the sort is disabled. But I expect the same interaction for all our list sections including that the create event (1) opens the sections, (2) opens the created item, (3) and focuses the first entry (which defines the list item name).
> 
> @MaxTru I would propose to discuss how we solve this. Maybe as an additional issue outside of this PR, but mandatory from my perspective.

Okay, if this is a must, we need to implement [this](https://github.com/bpmn-io/properties-panel/issues/96) first from my perspective. @pinussilvestrus would like to hear your opinion since you worked on this initially, do you agree? Thanks

philippfromme commented:
Last question: I changed the type of a field to `enum` and added a value. Then I changed the type to something else. The added value wasn't removed even though values are only used with type `enum`. In the old properties panel, values are removed.

![bsnV23iFJU](https://user-images.githubusercontent.com/7633572/133428569-e3c5728c-52c2-4f6c-bb1e-50358449e40f.gif)

Is this the intended behavior?

MaxTru commented:
> Last question: I changed the type of a field to `enum` and added a value. Then I changed the type to something else. The added value wasn't removed even though values are only used with type `enum`. In the old properties panel, values are removed.
> 
> ![bsnV23iFJU](https://user-images.githubusercontent.com/7633572/133428569-e3c5728c-52c2-4f6c-bb1e-50358449e40f.gif)
> 
> Is this the intended behavior?

Yes, this is kind of intended because we planned (see epic issue) to have a seperate behavior for that:
![Screenshot_20210915_135947](https://user-images.githubusercontent.com/42800119/133429499-10323c16-a945-412e-a353-21103fac7065.png)

However, after our last hoc discussion, I am not 100% confident this is the right approach.

Since you are the behavior-DRI, let's discuss face2face :-)

andreasgeier commented:
> Last question: I changed the type of a field to enum and added a value. Then I changed the type to something else. The added value wasn't removed even though values are only used with type enum. In the old properties panel, values are removed.
> Is this the intended behavior?

From the UI perspective, the general rule is that we remove everything and clean up the XML when switching a property that has related sub properties. We even changed old behavior to adapt to this rule (now, we clean up condition expressions in a connection flow which was not the case in the old panel).

MaxTru commented:
> > Last question: I changed the type of a field to enum and added a value. Then I changed the type to something else. The added value wasn't removed even though values are only used with type enum. In the old properties panel, values are removed.
> > Is this the intended behavior?
> 
> From the UI perspective, the general rule is that we remove everything and clean up the XML when switching a property that has related sub properties. We even changed old behavior to adapt to this rule (now, we clean up condition expressions in a connection flow which was not the case in the old panel).
> 
> @MaxTru @philippfromme pls include me if you want to disregard this rule.

100% => the clean-up will happen. This is more a technical discussion. Our approach is to have the "clean-up" as part of a seperate behavior. This ensures that this will also happen, if someone changes properties not via the panel but via API.

The only issue we have at the moment is that this behavior is not integrated when testing the properties-panel seperately. Philipp is aware of this problem, and is the DRI to think of a solution for this technical problem.

I hope this helps, if not please message me directly :-)

philippfromme commented:
Okay, in my opinion, we're ready to merge this PR.

1. Cleaning up `values` after changing the type to something other than `enum` will be done using a behavior
2. Auto-focusing new form fields will be made possible in `@bpmn-io/properties-panel` soon so we can use it for form fields as well

andreasgeier reviewed (approved):
According to the conditions discussed above, I can also approve the PR.

- clean up of sub properties such as enum values will happen, it will be implemented separately
- we'll find a solution for the expected UI behavior and solve this in another issue

Congratulations to @MaxTru for implementing this big thing
